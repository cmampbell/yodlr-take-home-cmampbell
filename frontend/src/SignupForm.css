.SignUpForm-div h1 {
    margin: 30px;
    text-transform: capitalize;
}

.SignUpForm input{
    background-color: #fcfcfc;
}

.SignUpForm button{
    background-color: #C7F5C4;
    color: black;
}

.SignUpForm button:hover{
    background-color: #88C2A5;
    color: black;
}