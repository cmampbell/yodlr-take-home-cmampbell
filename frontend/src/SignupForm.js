import React, { useState } from "react";
import { Form, useNavigate } from 'react-router-dom'
import { Input, FormGroup, Label, Button } from 'reactstrap'
import './SignupForm.css'

const SignupForm = () => {
    const navigate = useNavigate()

    const INITIAL_STATE = {
        firstName: '',
        lastName: '',
        email: '',
    };
    const [formData, setFormData] = useState(INITIAL_STATE);
    const [formValidator, setFormValidator] = useState(false)

    const validateField = () => {
        // we check the values in form data
        for (let key in formData) {
            if (!formData[key]) {
                // if any are empty strings set Validator to false
                setFormValidator(false)
                return
            }
        };
        setFormValidator(true)
    }

    const handleChange = (evt) => {
        validateField()
        const { name, value } = evt.target
        setFormData(() => ({
            ...formData, [name]: value
        }
        ))

    }

    const handleSubmit = async (evt) => {
        evt.preventDefault();
        await fetch(`http://localhost:3001/users`, {
            method: "POST",
            mode: "cors",
            headers: { "content-type": "application/json" },
            body: JSON.stringify(formData)
        })
        navigate('/')
    }

    return (
        <section className="SignUpForm-div">
            <h1>Sign up for Yodlr!</h1>
            <Form onSubmit={handleSubmit} className="SignUpForm">
                <FormGroup floating>
                    <Input
                        placeholder='first name'
                        name="firstName"
                        id="firstName"
                        type="text"
                        value={formData.firstName}
                        onChange={handleChange}
                    />
                    <Label htmlFor='firstName' className="label">First Name</Label>

                </FormGroup>
                <FormGroup floating>
                    <Input
                        placeholder='last name'
                        name="lastName"
                        id="lastName"
                        type="text"
                        value={formData.lastName}
                        onChange={handleChange}
                    />
                    <Label htmlFor='lastName'>Last Name</Label>

                </FormGroup>
                <FormGroup floating>
                    <Input
                        placeholder='email'
                        name="email"
                        id="email"
                        type="text"
                        value={formData.email}
                        onChange={handleChange}
                    />
                    <Label htmlFor='email'>Email</Label>

                </FormGroup>
                <Button type="submit" disabled={!formValidator}>Sign Up!</Button>
            </Form>
        </section>
    )
}

export default SignupForm;