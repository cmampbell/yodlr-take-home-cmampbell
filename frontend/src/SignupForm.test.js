import React from "react";
import renderWithRouter from "./renderWithRouter";
import SignupForm from "./SignupForm";
import { screen } from "@testing-library/react";
import userEvent from '@testing-library/user-event'
import { act } from "react-dom/test-utils";

const user = userEvent.setup()

describe('SignupForm tests', () => {
    it('should render without crashing', () => {
        // define a full route as the first argument
        renderWithRouter(<SignupForm />)
    })

    it('should match the snapshot', ()=> {
        const { asFragment } = renderWithRouter(<SignupForm />)

        expect(asFragment()).toMatchSnapshot();
    })

    it('should handle user input', async () => {
        renderWithRouter(<SignupForm />)

        await act(async ()=> {
            await user.type(screen.getByLabelText("First Name"), 'test');
            await user.type(screen.getByLabelText("Last Name"), 'test');
            await user.type(screen.getByLabelText("Email"), 'test@email.com');
            await user.click(screen.getByText("Sign Up!"))
        })


        expect(fetch).toHaveBeenCalled();
        expect(fetch).toHaveBeenCalledWith(`http://localhost:3001/users`, {
            method: "POST",
            mode: "cors",
            headers: { "content-type": "application/json" },
            body: JSON.stringify({
                firstName: 'test',
                lastName: 'test',
                email: 'test@email.com'
            })
        });
        
    })
})