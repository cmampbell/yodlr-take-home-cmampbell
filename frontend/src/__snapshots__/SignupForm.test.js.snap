// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupForm tests should match the snapshot 1`] = `
<DocumentFragment>
  <section
    class="SignUpForm-div"
  >
    <h1>
      Sign up for Yodlr!
    </h1>
    <form
      action="/"
      class="SignUpForm"
      method="get"
    >
      <div
        class="mb-3 form-floating"
      >
        <input
          class="form-control"
          id="firstName"
          name="firstName"
          placeholder="first name"
          type="text"
          value=""
        />
        <label
          class="label form-label"
          for="firstName"
        >
          First Name
        </label>
      </div>
      <div
        class="mb-3 form-floating"
      >
        <input
          class="form-control"
          id="lastName"
          name="lastName"
          placeholder="last name"
          type="text"
          value=""
        />
        <label
          class="form-label"
          for="lastName"
        >
          Last Name
        </label>
      </div>
      <div
        class="mb-3 form-floating"
      >
        <input
          class="form-control"
          id="email"
          name="email"
          placeholder="email"
          type="text"
          value=""
        />
        <label
          class="form-label"
          for="email"
        >
          Email
        </label>
      </div>
      <button
        class="btn btn-secondary disabled"
        disabled=""
        type="submit"
      >
        Sign Up!
      </button>
    </form>
  </section>
</DocumentFragment>
`;
